### PR TITLE
Fix SkiaLayer transform

### LINKED ISSUE #12: Modifier.drawLayer(rotationX=, rotationY=), support perspective projection
Now it is drawn with isometric projection. I.e. when we draw Box(Modifier.drawLayer(rotationX=45), it will just change its height.

May be it will help to implement perspective projection:
https://docs.microsoft.com/en-us/xamarin/xamarin-forms/user-interface/graphics/skiasharp/transforms/3d-rotation

Drawing of layers (and nested ones) should be similar to drawing of layers in Android.

### LINKED ISSUE #1129: SkijaLayer should support camera distance
Google has added support for camera distance APIs to provide skew visual effects. However, the current implementation of SkijaLayer does not expose Sk3dView within Skia to support the same API surface:

See:
https://android-review.googlesource.com/c/platform/frameworks/support/+/1475539/11/compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/platform/SkijaLayer.kt

### PR DESCRIPTION
**Pull request by MatkovIvan**
## Proposed Changes

Proper calculation of transformation matrix in `SkiaLayer`. Now it matches android implementation.

## Testing

Test: Compare result of applying `graphicsLayer` parameters (`rotation*` and `cameraDistance`) between android and desktop.

Please note that 2 tests are failing because golden images need to be updated. [TODO](https://github.com/JetBrains/compose-jb/issues/2819)

## Issues Fixed

Fixes https://github.com/JetBrains/compose-jb/issues/12
Fixes https://github.com/JetBrains/compose-jb/issues/1129

### CODE REVIEW

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaLayer.skiko.kt [hunk 1]
```diff
@@ -186,27 +192,26 @@ internal class SkiaLayer(
...
-            rotateY(rotationY)
             rotateZ(rotationZ)
+            rotateY(rotationY)
+            rotateX(rotationX)
             scale(scaleX, scaleY)
         }
         matrix *= Matrix().apply {
-            translate(x = pivotX, y = pivotY)
+            // the camera location is passed in inches, set in pt
+            val depth = cameraDistance * 72f
```
Comment (dima-avdeev-jb) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaLayer.skiko.kt:209]: As I understand:
depth - in pixels.
cameraDistance - in inches.
72 - in pixels per inch (PPI). But this multiplier should depend on screen resolution and size.

Or 72 is a universal value for such cases?

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaLayer.skiko.kt [hunk 2]
```diff
@@ -186,27 +192,26 @@ internal class SkiaLayer(
...
+            set(2, 3, -1f / depth)
+            set(2, 2, 0f)
```
Comment (dima-avdeev-jb) [inline compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaLayer.skiko.kt]:
```suggestion
            set(row = 2, column = 3, v = -1f / depth)
            set(row = 2, column = 2, v = 0f)
```
Maybe with argument names it will be more understandable?

File: compose/ui/ui/src/skikoTest/kotlin/androidx/compose/ui/platform/SkiaLayerTest.kt [hunk 3]
```diff
@@ -254,29 +259,31 @@ class SkiaLayerTest {
...
             rotationX = 45f,
-            transformOrigin = TransformOrigin(0f, 0f)
+            transformOrigin = TransformOrigin(0f, 0f),
+            cameraDistance = Float.MAX_VALUE
         )
         val matrix = layer.matrix
 
         val y = (10 * cos45).roundToInt()
-        val translationY = (7 * cos45).roundToInt()
+        val translationY = 7
```
Comment (dima-avdeev-jb) [inline compose/ui/ui/src/skikoTest/kotlin/androidx/compose/ui/platform/SkiaLayerTest.kt:268]: Why this changes?
It means that the previous value was a workaround to correct a wrong transformation?

**Review by dima-avdeev-jb — COMMENTED** [review]
Thanks! That's amazing math work.
Added some comments to clarify my understanding.

File: compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaLayer.skiko.kt (hunk 1 above)
Comment (dima-avdeev-jb) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaLayer.skiko.kt:209]: Or maybe depth - in Density independent pixels (dp)?

Comment (MatkovIvan) [reply compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/platform/SkiaLayer.skiko.kt:209]: The value and actually the comment picked up from [android implementation](https://cs.android.com/android/platform/superproject/+/master:external/skia/src/utils/SkCamera.cpp;l=181;drc=5ca657189aac546af0aafaba11bbc9c5d889eab3).

Initially, I thought that it should depend on density too. I event experimented with it in this branch (see commit history), but at the end - hardcoded `72` is the universal value to match android behavior.

File: compose/ui/ui/src/skikoTest/kotlin/androidx/compose/ui/platform/SkiaLayerTest.kt (hunk 3 above)
Comment (MatkovIvan) [reply compose/ui/ui/src/skikoTest/kotlin/androidx/compose/ui/platform/SkiaLayerTest.kt:268]: Yes. Note that `transformOrigin` is `(0, 0)`, so rotation around it should transfer this point.

**Review by dima-avdeev-jb — APPROVED** [review]

File: compose/ui/ui/src/skikoTest/kotlin/androidx/compose/ui/platform/SkiaLayerTest.kt [hunk 4]
```diff
@@ -254,29 +259,31 @@ class SkiaLayerTest {
             translationX = 60f,
             translationY = 7f,
             rotationX = 45f,
```
Comment (igordmn) [inline compose/ui/ui/src/skikoTest/kotlin/androidx/compose/ui/platform/SkiaLayerTest.kt:261]: Can we make tests that makes "trapeze" figure in the result screenshots? I.e. when we rotate by 45 degrees, the top should be narrower

**Review by igordmn — COMMENTED** [review]
Thanks!

**Review by igordmn — APPROVED** [review]
